auctions/dutch-auction/src/main.sw

```diff
@@ -0,0 +1,219 @@
+        storage.admin = admin;
+        storage.initialized = true;
+    }
+
+    /// Returns the current price for the auction corresponding to the auction_id
+    fn price(auction_id: u64) -> u64 {
+        validate_id(auction_id);
+
+        calculate_price(auction_id)
+    }
+
+    /// Bids in the given auction_id, is a win if the amount and asset are correct
+    fn bid(auction_id: u64) {
+        // In a Dutch auction the first bid wins
+        require(storage.initialized, TechnicalError::ContractNotYetInitialized);
+
+        validate_id(auction_id);
+
+        let mut auction = storage.auctions.get(auction_id);
+
+        require(!auction.ended, TimeError::AuctionAlreadyEnded);
+
+        // Cannot bid before auction starts
+        require(auction.start_time <= height(), TimeError::AuctionNotYetStarted);
+
+        // Checks for correct asset_id being sent
+        require(msg_asset_id() == auction.asset_id, BidError::WrongAssetSent);
+
+        let price = calculate_price(auction_id);
+
+        // Checks for high enough amount being sent
+        require(price <= msg_amount(), BidError::BidTooLow);
+
+        // Disallows furthur bids
+        auction.ended = true;
+        storage.auctions.insert(auction_id, auction);
+
+        // If someone sends more than the current price, refunds the extra amount
+        if msg_amount() > price {
+            let return_amount = msg_amount() - price;
+            transfer_to_identity(return_amount, auction.asset_id, sender_indentity());
+        }
+
+        on_win(auction, price);
+
+        log(WinningBidEvent {
+            winner: sender_indentity(), id: auction_id, 
+        });
+    }
+
+    /// Creates a new auction
+    fn create_auction(opening_price: u64, reserve_price: u64, start_time: u64, end_time: u64, beneficiary: Identity, asset: ContractId) {
+        require(storage.initialized, TechnicalError::ContractNotYetInitialized);
+        require(eq_identity(storage.admin, sender_indentity()), AuthorizationError::SenderNotAdmin);
+        require(reserve_price <= opening_price, SetupError::EndPriceCannotBeLargerThanStartPrice);
+        require(height() < end_time, SetupError::AuctionCannotEndInThePast);
+        require(height() <= start_time, SetupError::AuctionCannotStartInThePast);
+        require(start_time < end_time, SetupError::AuctionCannotEndBeforeItStarts);
+
+        storage.auction_count = storage.auction_count + 1;
+
+        let auction = Auction {
+            opening_price, reserve_price, start_time, end_time, beneficiary, asset_id: asset,
+            ended: false,
+        };
+
+        storage.auctions.insert(storage.auction_count, auction);
+
+        log(CreatedAuctionEvent {
+            id: storage.auction_count, auction, 
+        });
+    }
+
+    /// Cancels an auction so no one can bid on it. Needs to be called even after end_time of an auction if you want no one to bid anymore
```

Comment:
Split onto 2 lines. People set different character lengths for how long a line should be and I think they are typically 80, 100 & 120 characters long.
I use 100 iirc. Set your editor to draw a line to indicate the character limit so that you do not write lines that are too long.

Reply:
"Needs to be called even after end_time of an auction if you want no one to bid anymore" needs to be rephrased.